knative-operator/pkg/common/telemetry/telemetry.go

```diff
@@ -0,0 +1,180 @@
+package telemetry
+
+import (
+	"context"
+	"os"
+	"sync"
+
+	"github.com/openshift-knative/serverless-operator/knative-operator/pkg/common"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	kafkasourcev1beta1 "knative.dev/eventing-contrib/kafka/source/pkg/apis/sources/v1beta1"
+	eventingsourcesv1beta1 "knative.dev/eventing/pkg/apis/sources/v1beta1"
+	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const SkipTelemetryEnvVar = "SKIP_TELEMETRY"
+
+var (
+	log             = common.Log.WithName("telemetry")
+	metricsSnapshot = map[string]*sync.Map{
+		"eventing":     {},
+		"serving":      {},
+		"knativeKafka": {},
+	}
+)
+
+// TryStartTelemetry setups telemetry per component either Eventing, KnativeKafka or Serving.
+// When called it assumes that the component has status ready.
+func TryStartTelemetry(c controller.Controller, mgr manager.Manager, stop chan struct{}, component string, api client.Client) error {
```

Comment:
I would've expected this to be part of a `Telemetry` type that you could store on the reconciler itself. You'd then just call `Start` and `Stop` accordingly.

This would solve two issues in the current code:
1. The pointer-replacement shenanigans with the stop channel below.
2. The skip env var becomes unnecessary as you can initialize the reconciler with a noop telemetry object in the test.

Reply:
@markusthoemmes I can try that to improve compaction but I dont see them as issues. The reason I followed this is because I didnt want to add a lot of noise in the reconciler anyway.

Reply:
True, they are not issues per-se, but they are unnecessary obstructions imo. Types are cheap. I think this currently adds more noise to the reconcilers than encapsulating the telemetry controllers in a type would.